### ISSUE TITLE
Aporeto connectivity issue with Patroni | SPIKE

### ISSUE THREAD

**Issue by StevenBarre** [BCDevOps/developer-experience#834]
**TL;DR**
Long running pods may have their Aporeto Processing Unit (backend object) removed and re-added randomly. Without a matching PU traffic from/to the pod will be rejected. This has mostly been seen with pods that connect to a database, but may not be limited to that.

**Describe the issue**
![Clipboard - January 7, 2021 11_42 AM.png](https://images.zenhubusercontent.com/5d544a7a5faedc627998cd85/012f1820-054b-4c55-a025-3afd7f8e164d)

Each morning one of the pods goes from successfully connecting to the Patroni service to failing to connect. 

**Additional context**
Namespace: 264e6f-prod
User: aidan.cope

Namespace: df1ee0-prod
User: karim.gillani

**Definition of done**
No more disconnects

**Comment by sbarre-esit** [comment]
Working with Julian from Aporeto on investigating.

**Comment by ikethecoder** [comment]
So a different bit of odd behavior w/r to Apo... last night at about 10:27pm I deployed a new Pod to our prod environment and I did not setup any NSP, so as expected traffic was not flowing to it from another source Pod.  But at about 8:27am this morning, traffic suddenly opened up between the two Pods until 4:43p, when it then blocked traffic again.  At which time I was alerted and I updated the NSP and traffic is now correctly communicating between the two Pods.  But would like to know why from 8:27a to 4:43p the protection was not there?
Also whether it has any correlation with the intermittent issue of connectivity stopping between Pods without reason.

**Comment by sbarre-esit** [comment]
@ikethecoder what namespace was this in and what was the name of the pod?

**Comment by sbarre-esit** [comment]
Aporeto says:
> I've opened a blocker level ticket, waiting to hear from engineering. We will likely ask to take a closer look at the problem.

**Comment by ikethecoder** [comment]
Namespace was 264e6f-prod and a Pod was kong-kong-b86d69c58-d4c9n

**Comment by ikethecoder** [comment]
Another timeout issue, this was Kong Pod (kong-kong-b86d69c58-25gw7) connection to Patroni (patroni-spilo-1) and FluentD Aggregator (fluentd-0).
Started blocking with connection timeout at 11:33pm (Jan 13) and connections magically opened up again at 6:31am (Jan 14).

**Comment by sbarre-esit** [comment]
Hmm, for `kong-kong-b86d69c58-25gw7` looks like Aporeto removed the PU, and then added it back later. We'll add this to the  debug info.

```
# oc logs enforcerd-worker-crlsg | grep be260941-6459-40a5-9a7f-a42ac1ca272c
{"l":"info","t":1610475043.3699658,"m":"Using compressed tags","srv":"be260941-6459-40a5-9a7f-a42ac1ca272c","version":"\u0001"}
{"l":"info","t":1610475043.3780317,"m":"Enabling Trireme Datapath v2.0","srv":"be260941-6459-40a5-9a7f-a42ac1ca272c"}
{"l":"info","t":1610475043.3780537,"m":"eBPF is disabled as it is not supported","srv":"be260941-6459-40a5-9a7f-a42ac1ca272c"}
{"l":"info","t":1610475043.4918754,"c":"policy/engine.go:404","m":"Processing unit started","type":"Docker","name":"kong-kong-b86d69c58-25gw7","id":"5ffde61ac3a26d0001a3a2fa","nativeID":"be260941-6459-40a5-9a7f-a42ac1ca272c","status":"Running","namespace":"/bcgov/platform-services/production/kamloops/silver/264e6f-prod"}
{"l":"info","t":1610609135.6740553,"m":"Remote enforcer exiting ...","srv":"be260941-6459-40a5-9a7f-a42ac1ca272c"}
{"l":"info","t":1610609135.8172057,"c":"policy/engine.go:639","m":"Processing unit removed","type":"Docker","name":"kong-kong-b86d69c58-25gw7","id":"5ffde61ac3a26d0001a3a2fa","nativeID":"be260941-6459-40a5-9a7f-a42ac1ca272c","status":"Terminated"}
{"l":"info","t":1610634356.8422792,"m":"Using compressed tags","srv":"be260941-6459-40a5-9a7f-a42ac1ca272c","version":"\u0001"}
{"l":"info","t":1610634356.851823,"m":"Enabling Trireme Datapath v2.0","srv":"be260941-6459-40a5-9a7f-a42ac1ca272c"}
{"l":"info","t":1610634356.8518553,"m":"eBPF is disabled as it is not supported","srv":"be260941-6459-40a5-9a7f-a42ac1ca272c"}
{"l":"info","t":1610634356.9721322,"c":"policy/engine.go:404","m":"Processing unit started","type":"Docker","name":"kong-kong-b86d69c58-25gw7","id":"60005474c3a26d0001a3f127","nativeID":"be260941-6459-40a5-9a7f-a42ac1ca272c","status":"Running","namespace":"/bcgov/platform-services/production/kamloops/silver/264e6f-prod"}
```

**Comment by j-pye** [comment]
Meeting Notes - Jan 14th - 1PM PT

Aporeto is actively working on a couple of fixes. One is complete and the other is nearing completion. They expect to cut a release for a new enforcer image this afternoon or tomorrow. We expect to roll out the new enforcer (with a rolling drain) tomorrow/Monday that should fix the issue. It will go through KLAB first to make sure it gets tested in a dev environment before Silver is touched.

/cc @mitovskaol @juhewitt

**Comment by ikethecoder** [comment]
Was Aporeto able to provide any possible workaround?   Such as having a more permissive NSP like allowing all inter-pod communication within a namespace?

**Comment by sbarre-esit** [comment]
We haven't found any workarounds yet unfortunately.

**Comment by jefkel** [comment]
If you have added external dependent service checks to your readinessCheck for your pods, that will mitigate the effects of the issue (as well as make your service more resilient)

An example would be to add a check for database connectivity to your pod Readiness check (checking if db connection status available/rejected/timeout)

If the DB is not available (rejected or timeout status), then your pod can be marked "notReady" and the service will no longer forward traffic to the pod that's having issues.

The current Aporeto issue we're seeing is intermittently creating a timeout scenario for a pod, and a readinessCheck that verifies your DB connectivity would mitigate the service impacts.

**Comment by sbarre-esit** [comment]
Aporeto says:

>Release candidate for the fix is being tested right now. Tests should be done by tomorrow morning.
We will have a patched version tomorrow 01/20/2021 afternoon.
The official 3.14.9 release will be cut after all other testing is done, which will be around Friday 01/22/2021.

**Comment by sbarre-esit** [comment]
Aporeto says:

>Unfortunately, I have bad news about the availability of a fix for the deleted PU problem. 
>
>QA was not able to fully reproduce the problem and test the solution, meaning there is additional bug that needs to be fixed. This will push the bugfix release to next week.

**Comment by sbarre-esit** [comment]
Aporeto says:
>Additional fixes needed to mitigate PU removal bug were merged today. The QA cycle was started. New release 3.14.9 will be available by EOD Thursday 01/28/2021.

**Comment by sbarre-esit** [comment]
Aporeto says
>We are about to cut the release, running a few hours late. Worst case scenario, 3.14.9 should be out on Monday

Once released Platform Ops will install in one of our labs to validate and then communicate the plan/date for SILVER to be upgraded.

**Comment by StevenBarre** [comment]
Aporeto 3.14.9 has been released and we are beginning testing in the LABS. We're aiming to roll it out to SILVER Monday Feb 8th.

**Comment by ikethecoder** [comment]
Is this now a wash due to the decision to move to OCP4 network policies?

**Comment by StevenBarre** [comment]
@ikethecoder No, because Aporeto will be around until March 31st while people transition. So improvements to stability, even for 2 months, is worth doing.

**Comment by StevenBarre** [comment]
CHG0022904 MCS SILVER - Aporeto Enforcer upgrade

has been complete. Will close this ticket in a few days if nothing else comes up.

**Comment by StevenBarre** [comment]
So far things look good on the new version. Closing this ticket.
